state-chain/pallets/cf-environment/src/tests.rs

```diff
@@ -226,3 +153,222 @@ fn update_safe_mode() {
+			consolidation_threshold: 5,
+			consolidation_size: 2,
+		});
+
+		BitcoinAvailableUtxos::<Test>::set(vec![
+			utxo(1_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(2_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(3_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(31_000_000, CHANGE_ADDRESS_SALT, Some(epoch_3)),
+			utxo(32_000_000, CHANGE_ADDRESS_SALT, Some(epoch_3)),
+			utxo(33_000_000, CHANGE_ADDRESS_SALT, Some(epoch_3)),
+		]);
+
+		assert_eq!(
+			Environment::select_and_take_bitcoin_utxos(UtxoSelectionType::SelectForConsolidation),
+			None
+		);
+
+		// No consolidation. Epoch 1 utxos are discarded
+		CurrentBitcoinKey::set(Some(agg_key(epoch_3, Some(epoch_2))));
+		assert_eq!(
+			Environment::select_and_take_bitcoin_utxos(UtxoSelectionType::SelectForConsolidation),
+			None
+		);
+
+		System::assert_has_event(RuntimeEvent::Environment(crate::Event::StaleUtxosDiscarded {
+			utxos: vec![
+				utxo(1_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+				utxo(2_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+				utxo(3_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			],
+		}));
+
+		// Can consolidate and discard at the same time
+		BitcoinAvailableUtxos::<Test>::append(utxo(1_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)));
+
+		CurrentBitcoinKey::set(Some(agg_key(epoch_4, Some(epoch_3))));
+		assert!(Environment::select_and_take_bitcoin_utxos(
+			UtxoSelectionType::SelectForConsolidation
+		)
+		.is_some());
+		System::assert_has_event(RuntimeEvent::Environment(crate::Event::StaleUtxosDiscarded {
+			utxos: vec![utxo(1_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1))],
+		}));
+	});
+}
+
+#[test]
+fn can_consolidate_current_and_prev_utxos() {
+	let epoch_1 = [0xAA; 32];
+	let epoch_2 = [0xBB; 32];
+
+	new_test_ext().execute_with(|| {
+		CurrentBitcoinKey::set(Some(agg_key(epoch_2, Some(epoch_1))));
+		ConsolidationParameters::<Test>::set(utxo_selection::ConsolidationParameters {
+			consolidation_threshold: 5,
+			consolidation_size: 4,
+		});
+
+		BitcoinAvailableUtxos::<Test>::set(vec![
+			utxo(11_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(12_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(13_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+			utxo(21_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+			utxo(22_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+			utxo(23_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+			utxo(24_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+			utxo(25_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+		]);
+
+		// Consolidate from storage. Take the first 4 utxos.
+		assert_eq!(
+			Environment::select_and_take_bitcoin_utxos(UtxoSelectionType::SelectForConsolidation),
+			Environment::calculate_utxos_and_change(vec![
+				utxo(11_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+				utxo(12_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+				utxo(13_000_000, CHANGE_ADDRESS_SALT, Some(epoch_1)),
+				utxo(21_000_000, CHANGE_ADDRESS_SALT, Some(epoch_2)),
+			])
+		);
```

Comment:
Another example: the `4` could be a constant, and then we can just assert that the output contains the expected number of utxos. This test should not be concerned with anything else. 

We can write a separate test for `calculate_utxos_and_change` to make sure it does so correctly.